### DISCUSSION TITLE
Flipping port order when reading touchstone files

### DISCUSSION THREAD

**Discussion by kcbrowning** [scikit-rf/scikit-rf#823] [Q&A] [answered]
When the touchstone files are converted to freq x n x n matrix, the port order/mapping is changed as shown below.
skrf.Network has the same issue. 

```
Python 3.8.7 (tags/v3.8.7:6503f05, Dec 21 2020, 17:59:51) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import skrf
>>> import numpy as np
>>> skrf.__version__
'0.24.1'
>>> np.__version__
'1.24.1'
>>> touch = skrf.Touchstone('SN-0000000719_0.022.s2p')
>>> touch.get_sparameter_names()
['frequency', 'S11R', 'S11I', 'S12R', 'S12I', 'S21R', 'S21I', 'S22R', 'S22I']
>>> touch.sparameters[0, :]      
array([ 1.40000e+10,  1.81182e-02,  6.81423e-02, -5.29483e-05,
        4.49923e-05,  5.74469e-01,  8.08992e-01, -1.14410e-01,
        9.91022e-02])
>>> [k for k in touch.get_sparameter_data().keys()]
['frequency', 'S11R', 'S11I', 'S12R', 'S12I', 'S21R', 'S21I', 'S22R', 'S22I']
>>> np.array([v[0] for v in touch.get_sparameter_data().values()])
array([ 1.40000e+10,  1.81182e-02,  6.81423e-02,  5.74469e-01,
        8.08992e-01, -5.29483e-05,  4.49923e-05, -1.14410e-01,
        9.91022e-02])
>>> [(k, v[0]) for k, v in touch.get_sparameter_data().items()] 
[('frequency', 14000000000.0), ('S11R', 0.0181182), ('S11I', 0.0681423), ('S12R', 0.574469), ('S12I', 0.808992), ('S21R', -5.29483e-05), ('S21I', 4.49923e-05), ('S22R', -0.11441), ('S22I', 0.0991022)]
>>> [(k, v) for k, v in zip(touch.get_sparameter_names(), touch.sparameters[0, :])]    
[('frequency', 14000000000.0), ('S11R', 0.0181182), ('S11I', 0.0681423), ('S12R', -5.29483e-05), ('S12I', 4.49923e-05), ('S21R', 0.574469), ('S21I', 0.808992), ('S22R', -0.11441), ('S22I', 0.0991022)]
```

**Comment by Ttl** [comment, marked as answer]
For historical reasons two-port Touchstone S-parameters are in the transposed order compared to N-port S-parameters. The low level `sparameters` and `get_sparameter_names` methods return the original data in the original transposed order but the names follow the general N-port pattern. The data is correctly transposed when converted to matrix. The issue is with `get_sparameter_names` not handling the two-port special case. 

`skrf.Network('file.s2p')` is the recommended method for reading Touchstone files.

**Comment by kcbrowning** [comment]
`skrf.Network('file.s2p')` does the same thing or maybe I am misunderstanding the index.
Touchstone file contains this order
```
!S2P File: Measurements: S11, S12, S21, S22:
# Hz S RI R 50
14000000000.000000 0.0181182 0.0681423 -5.29483e-05 4.49923e-05 0.574469 0.808992 -0.11441 0.0991022
```
```
>>> ntwk = skrf.Network('file.s2p')
>>> ntwk.s[0, :, :]
array([[ 1.81182e-02+6.81423e-02j  5.74469e-01+8.08992e-01j]
          [-5.29483e-05+4.49923e-05j -1.14410e-01+9.91022e-02j]])
```
I assumed that `skrf.network.s[freq, i, j]` meant S(i+1)(j+1), but it appears to mean S(j+1)(i+1).
The documentation for `skrf.network.s` only provides an example for S11 (`skrf.network.s[:,0,0]`). If I want S21, my initial thought was `skrf.network.s[:,1,0]`, but your saying it should be `skrf.network.s[:,0,1]`. So, if I had a s3p file, S23 would be `skrf.network.s[:,2,1]`? Or is this just a special case for s2p?

**Comment by Ttl** [comment]
You assumed correctly about the S-parameter ordering. Touchstone file format specifies that for two-port file the order of parameters is `S11 S21 S12 S22` (https://ibis.org/connector/touchstone_spec11.pdf). N-port should be in `S11 S12 S13 S14 ...` format. This order is not verified in scikit-rf and it's assumed that the input file is a valid Touchstone file following the specified order.

**Comment by kcbrowning** [comment]
Well I learned something today and my custom touchstone writer needs to have a special case for ordering 2 port files.
Thanks
